_drafts/gnu-build-id-for-firmware.md

```diff
@@ -0,0 +1,270 @@
+
+The build ID is a 160-bit SHA1 string computed over the elf header bits and
+section contents in the file. It is bundled in the elf file as an entry in the
+notes section.
+
+Each note section entry has the following layout:
+
+```
++----------------+
+|     namesz     |   32-bit, size of "name" field
++----------------+
+|     descsz     |   32-bit, size of "desc" field
++----------------+
+|      type      |   32-bit, vendor specific "type"
++----------------+
+|      name      |   "namesz" bytes, null-terminated string
++----------------+
+|      desc      |   "descsz" bytes, binary data
++----------------+
+
+```
+
+In the case of the GNU build ID:
+* `name` is `"GNU\0"`, which gives us `namesz` = 4
+* `desc` is our 160-bit SHA1 value, which gives us `descsz` = 20
+* `type` is 3
+
+## Adding the GNU build ID to your builds
+
+**Note**: all of our example are based on the 
+[minimal](https://github.com/memfault/zero-to-main/tree/master/minimal) program from our Zero to main()
+series. 
+
+In GCC, you can enable build IDs with the `-Wl,--build-id` which passes the
+`--build-id` flag to the linker. You can then read it back by dumping the notes
+section of the resulting elf file with `readelf -n`.
+
+By default, this is not enabled:
+
+```terminal
+francois-mba:minimal francois$ make clean all
+...
+arm-none-eabi-size build/minimal.elf
+   text    data     bss     dec     hex filename
+   1252       0    8192    9444    24e4 build/minimal.elf
+...
+francois-mba:minimal francois$ arm-none-eabi-readelf -n build/minimal.elf
+[No output]
+```
+
+But a small change to the CFLAGS is all it takes:
+```terminal
+francois-mba:minimal francois$ CFLAGS="-Wl,--build-id" make clean all
+build/minimal.elf
+...
+arm-none-eabi-size build/minimal.elf
+   text    data     bss     dec     hex filename
+   1288       0    8192    9480    2508 build/minimal.elf
+francois-mba:minimal francois$ arm-none-eabi-readelf -n build/minimal.elf
+
+Displaying notes found in: .note.gnu.build-id
+  Owner                 Data size       Description
+  GNU                  0x00000014       NT_GNU_BUILD_ID (unique build ID
+bitstring)
+    Build ID: bab6b09f86b3c3017499d8e386447a610c559bd5
+```
+
+As you can see, our binary now contains the build id
+`bab6b09f86b3c3017499d8e386447a610c559bd5`.
+
+## Bundling the GNU build ID in firmware bin files
+
+Getting the build ID in your executables on Linux is easy: they are ELF files!
+Firmware on the other hand typically deals with binaries which are assembling by
+copying relevant sections of the elf at the right offset in a file.
+
+This is typically accomplished with objcopy:
+```terminal
+$ arm-none-eabi-objcopy firmware.elf firmware.bin -O binary
+```
+
+This takes every elf section earmarked to be loaded and places them at 
+the correct offset in the bin file. In the process, most debug sections are stripped out.
+
+Dumping the elf sections of the resulting `minimal.elf` gives us:
```

Comment:
Maybe include what tool /CLI command you used to dump. This looks like `objdump -h`?